kodi_addon_checker/check_addon_branches.py

```diff
@@ -0,0 +1,116 @@
+"""
+    Copyright (C) 2017-2018 Team Kodi
+    This file is part of Kodi - kodi.tv
+
+    SPDX-License-Identifier: GPL-3.0-only
+    See LICENSES/README.md for more information.
+"""
+
+import logging
+import os
+import xml.etree.ElementTree as ET
+
+from .record import INFORMATION, PROBLEM, Record
+from .report import Report
+from .versions import AddonVersion, KodiVersion
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def check_for_existing_addon(report: Report, addon_path: str, all_repo_addons: dict, pr: bool,
+                             kodi_version: KodiVersion):
+    """Check if addon submitted already exists or not
+        :addon_path: path of the addon
+        :all_repo_addons: dictionary return by all_repo_addon() function
+    """
+
+    addon_xml = os.path.join(addon_path, "addon.xml")
+    addon_name, addon_version = _get_addon_name(addon_xml)
+    addon_details = {'name': addon_name, 'version': addon_version}
+
+    is_new_addon = True
+    for branch, repo in sorted(all_repo_addons.items(), reverse=True):
+
+        # Addon submission must be higher than the versions already available in lower
+        # branches to allow users to receive the update
+        if KodiVersion(branch) <= kodi_version and addon_name in repo:
+            is_new_addon = False
+            _check_version_higher(report, addon_details, branch, repo.find(addon_name).version, pr)
+
+        # Addon submission must be lower than the versions already available in upper branches
```

Comment:
Did this change? Normally it worked to push a higher version to a lower repo and have it overwrite all repos above too? Or is that due to the py3 changes?

Reply:
And yes, I know this is confusing for the repo maintainers.

Reply:
Yes that is the main change (althought there are others mainly code simplification). This isn't only related to py3. If you update a version for your addon in a lower branch (let's say gotham) and have the same add-on on an upper branch (eg Leia), the version you assign to the add-on in Leia must be lower than the version of the add-on available in Leia. If they are not, a user migrating from gotham to Leia will not receive the update (lower version). If an add-on exist in both repos, theoretically they are not compatible (python api's, language stuff, etc)

Reply:
I think you are right and this needs further work